tuxemon/db.py

```diff
@@ -725,7 +728,7 @@ def lookup(
     ) -> EnvironmentModel:
         pass
 
-    def lookup(self, slug: str, table: str = "monster") -> Mapping[str, Any]:
+    def lookup(self, slug, table="monster"):
```

Comment:
Why did you remove the return value?

Reply:
Just read the article more carefully... typehints are optional on the implementation. We may as well keep them.

Reply:
```
    @overload
    def lookup(
        self,
        slug: str,
        table: Literal["environment"],
    ) -> EnvironmentModel:
        pass

    @overload
    def lookup(
        self,
        slug: str,
        table: str,
    ) -> Mapping[str, Any]:
        pass
```
Sigh... mypy doesn't understand this, even though this behaviour is explained in the PEP https://peps.python.org/pep-0586/#interactions-with-overloads

Reply:
Simple reproduction of the mypy error:

```
from typing import Literal, Union, overload


@overload
def foo(s: Literal["hi"]) -> int:
    ...


@overload
def foo(s: str) -> str:
    ...


def foo(s: str) -> Union[str, int]:
    if s == "hi":
        return 1
    return s
```
example.py:5: error: Overloaded function signatures 1 and 2 overlap with incompatible return types

___

Will skip this one for now.

Reply:
Actually, I see what's going on here. The fallback option needs to be a superset of the literals, there's no option for "string that isn't one of the previous literals"

Reply:
The REAL problem is that since  #1192 landed, the returning type is not a `Mapping[str, Any]`, but a pydantic `BaseModel`. That is the "incompatible return types" Mypy complains about.

Reply:
Not quite, there's still one branch where it is a Mapping[str, Any]

https://github.com/Tuxemon/Tuxemon/blob/1c9b59686679a4ca7685027292e0881d1545a55d/tuxemon/db.py#L682

Reply:
That should be fixed then, to have a common base class.

Reply:
Is the common base class just for the existing models, or do you want the Mapping[str, Any] to be included somehow?

I guess for unknown tables we could wrap the dict in a Model. We shouldn't have unknown tables though, so don't see much point in keeping this around. I guess it could be useful for development when you're implementing a new table and want to use a dict at first, rather than defining the model straight away.

```
class NewModel(BaseModel):
    data: Mapping[str, Any]
```

tbh, I don't really like how the lookup function takes a table argument at all.

It should just be something like:

db.monsters.get(slug)

monsters could be a property if necessary